File: python/ch02_agent/agent_ch02/bootstrap.py
"""组合根：把外部实现装配为第 2 章 Agent。

这是什么：依赖注入的组装层，负责创建和连接所有依赖
Java 类比：类似 Spring 的 @Configuration 类或手动依赖注入的工厂类
为什么需要：集中管理对象创建和依赖关系，避免核心逻辑直接 new 具体实现

Java 对照：可以把本文件理解为 Spring 的 `@Configuration` 类。
它负责创建对象和连接依赖，但不负责真正的 Agent 业务流程。
"""

from .adapters.filesystem import LocalWorkspaceFileSystem
from .adapters.powershell import PowerShellRunner
from .core.commands import CommandRunner
from .core.filesystem import WorkspaceFileSystem
from .core.loop import AgentRunner, ToolAuthorizer
from .core.model import ModelClient
from .core.profiles import P01, P02, ChapterProfile
from .features.builtin_tools import create_chapter_one_tools, create_chapter_two_tools

SYSTEM_PROMPT = "You are a coding agent. Use tools when needed, inspect their results, and answer accurately."
# Java 对照：上面是模块常量，作用近似 `private static final String SYSTEM_PROMPT`。


def build_agent(profile: ChapterProfile, model: ModelClient, workspace: str, command_runner: CommandRunner | None = None, file_system: WorkspaceFileSystem | None = None, authorizer: ToolAuthorizer | None = None, max_turns: int = 20) -> AgentRunner:
    """创建一个第 2 章 AgentRunner。

    这是什么：Agent 工厂方法，根据章节配置组装完整的 Agent
    Java 类比：类似 @Bean AgentRunner buildAgent(...) { return new AgentRunner(...); }
    为什么需要：根据不同章节配置选择工具集，支持测试时注入 Fake 依赖

    `command_runner or PowerShellRunner()` 表示：调用者传了 Fake 就用 Fake，
    没传时才创建真实 PowerShellRunner。这就是最简单的依赖注入。
    """
    # `is` 比较对象身份，不是字段值。这样调用方不能 new 一个内容相同的 DTO
    # 冒充固定章节配置，Java 中相当于只接受预定义的单例常量。
    if profile is not P01 and profile is not P02:  # 只接受预定义的章节配置常量
        raise ValueError("必须传入固定的章节配置对象")
    command = command_runner or PowerShellRunner()  # 使用传入的 runner 或创建默认实例
    tools = (  # 根据章节选择工具集
        create_chapter_one_tools(command)
        if profile.chapter == 1  # 第 1 章只有命令工具
        else create_chapter_two_tools(command, file_system or LocalWorkspaceFileSystem())  # 第 2 章增加文件工具
    )
    return AgentRunner(model, tools, SYSTEM_PROMPT, workspace, max_turns=max_turns, authorizer=authorizer)  # 组装并返回 Agent

File: typescript/chapters/ch02/src/bootstrap.ts
/**
 * 组装层：按固定 ChapterProfile 选择 P01/P02 工具集并创建 AgentRunner。
 * 依赖都可注入，离线测试无需真实模型、进程或文件系统。
 * 工作区边界由 adapters 执行，这里只负责组合 profile、工具集、授权器和最大轮次。
 */
import { NodeWorkspaceFileSystem } from "./adapters/filesystem.js";
import { PowerShellRunner } from "./adapters/powershell.js";
import type { CommandRunner } from "./core/commands.js";
import type { WorkspaceFileSystem } from "./core/filesystem.js";
import { AgentRunner } from "./core/loop.js";
import type { ToolAuthorizer } from "./core/loop.js";
import type { ModelClient } from "./core/model.js";
import type { ChapterProfile } from "./core/profiles.js";
import { profileForChapter } from "./core/profiles.js";
import { createChapterOneTools, createChapterTwoTools } from "./features/builtin-tools.js";

// 组装层仅负责按章节档案选择基础设施，核心循环不依赖具体 SDK 或 Node I/O 实现。
// 第 2 章在第 1 章循环上注入受工作区约束的文件系统能力。
const SYSTEM_PROMPT =
  "You are a coding agent. Use tools when needed, inspect their results, and answer accurately.";

export interface BuildDependencies {
  // 可替换依赖使离线测试无需真实模型、进程或文件系统。
  readonly model: ModelClient;
  // 工具可访问的工作区根；具体路径安全检查由文件适配器执行。
  readonly workspace: string;
  // 可选命令进程边界，缺失时构造真实 PowerShellRunner。
  readonly commandRunner?: CommandRunner;
  // 可选文件系统边界，缺失时构造 NodeWorkspaceFileSystem。
  readonly fileSystem?: WorkspaceFileSystem;
  // 可选副作用审批边界。
  readonly authorizer?: ToolAuthorizer;
  // 可选模型请求次数上限。
  readonly maxTurns?: number;
}

// 根据固定 profile 装配累计工具集；拒绝伪造 profile 防止章节能力越级。
export function buildAgent(profile: ChapterProfile, dependencies: BuildDependencies): AgentRunner {
  // 只接受固定 profile，随后按章节选择累积工具集。
  if (profileForChapter(profile.chapter) !== profile) {
    throw new Error("profile must be a fixed chapter profile");
  }
  const commandRunner =
    dependencies.commandRunner === undefined ? new PowerShellRunner() : dependencies.commandRunner;
  const tools =
    // P01 只暴露 Shell；P02 才能取得文件工具，防止章节能力提前泄漏。
    profile.chapter === 1
      ? createChapterOneTools(commandRunner)
      : createChapterTwoTools(
          commandRunner,
          dependencies.fileSystem === undefined
            ? new NodeWorkspaceFileSystem()
            : dependencies.fileSystem,
        );
  return new AgentRunner({
    model: dependencies.model,
    tools,
    systemPrompt: SYSTEM_PROMPT,
    workspace: dependencies.workspace,
    ...(dependencies.authorizer === undefined ? {} : { authorizer: dependencies.authorizer }),
    ...(dependencies.maxTurns === undefined ? {} : { maxTurns: dependencies.maxTurns }),
  });
}
